ggml WebGPU: remove userdata from request adapter callback

danbev opened the pull request:
This commit removes the `userdata` parameter from the WebGPU request adapter callback in `ggml-webgpu.cpp`. Instead, the lambda function captures the `webgpu_context` directly.

The motivation for this change is to simplify the code and improve readability.

danbev committed:
ggml WebGPU: remove userdata from request adapter callback

This commit removes the `userdata` parameter from the WebGPU request
adapter callback in `ggml-webgpu.cpp`. Instead, the lambda function
captures the `webgpu_context` directly.

The motivation for this change is to simplify the code and improve
readability.

ngxson reviewed (commented):
I prefer keeping `void * userdata` because:
- It's possible to make the `callback` a `static` function
- More crucially, it maintains the correct callback function signature defined [here](https://eliemichel.github.io/LearnWebGPU/getting-started/adapter-and-device/the-adapter.html#asynchronous-function)

danbev commented:
> I prefer keeping `void * userdata` because:
> 
> * It's possible to make the `callback` a `static` function
> * More crucially, it maintains the correct callback function signature defined [here](https://eliemichel.github.io/LearnWebGPU/getting-started/adapter-and-device/the-adapter.html#asynchronous-function)

I was just thinking that since the implementation is using [webgpu/webgpu_cpp.h](https://github.com/ggml-org/llama.cpp/blob/34bdbbd7c2b70b848718e95bc48010f6aecd2816/ggml/src/ggml-webgpu/ggml-webgpu.cpp#L12) it might be a nice change in favor of somewhat simpler/more readable code (at least in my opinion).

ngxson commented:
Hmm ok I didn't notice that webgpu_cpp supports both version of callback (w/ and w/o `userdata`)

Probably it's also cleaner to move the callback to the same line with function call, like in this example: https://eliemichel.github.io/LearnWebGPU/getting-started/cpp-idioms.html#capturing-closures

danbev committed:
inline the callback lambda into the RequestAdapter call

This commit removes the callback lambda variable and inlines it directly
into the RequestAdapter call.

ngxson reviewed (approved):
seems like a good change, is there anything blocking this from being merged?

danbev commented:
> seems like a good change, is there anything blocking this from being merged?

Nothing from my end :+1:
